File: archivelib-sys2/c-lib/src/r_compress/fn207.cpp

#include "r_compress.hpp"

void RCompress::fn207() {
  uint32_t _226, _289, _229, _454, _455;
  uint32_t _456 = 0;
  uint16_t _217[2 * CONST_N145_IS_19 - 1];
  _229 = fn211(CONST_N141_IS_511, data->dat_arr191, data->dat_arr180,
               data->dat_arr192);
  _455 = data->dat_arr191[_229];
  write_bits_to_buffer(16, (uint16_t)_455);
  if (_229 >= CONST_N141_IS_511) {
    fn216(_217);
    _229 = fn211(CONST_N145_IS_19, _217, data->dat_arr181, data->dat_arr194);
    if (_229 >= CONST_N145_IS_19) {
      fn218(CONST_N145_IS_19, CONST_N147_IS_5, 3);
    } else {
      write_bits_to_buffer(CONST_N147_IS_5, 0);
      write_bits_to_buffer(CONST_N147_IS_5, (uint16_t)_229);
    }
    fn222();
  } else {
    write_bits_to_buffer(CONST_N147_IS_5, 0);
    write_bits_to_buffer(CONST_N147_IS_5, 0);
    write_bits_to_buffer(CONST_N143_IS_9, 0);
    write_bits_to_buffer(CONST_N143_IS_9, (uint16_t)_229);
  }
  _229 = fn211(CONST_N142_IS_15, data->dat_arr193, data->dat_arr181,
               data->dat_arr194);
  if (_229 >= CONST_N142_IS_15) {
    fn218(CONST_N142_IS_15, CONST_N540_IS_5, -1);
  } else {
    write_bits_to_buffer(CONST_N540_IS_5, 0);
    write_bits_to_buffer(CONST_N540_IS_5, (uint16_t)_229);
  }
  _454 = 0;
  for (_226 = 0; _226 < _455; _226++) {
    if (_226 % CHAR_BIT == 0)
      _456 = data->dat_arr165[_454++];
    else
      _456 <<= 1;
    if (_456 & (1U << (CHAR_BIT - 1))) {
      write_stored_bits_to_buffer(
          (int16_t)(data->dat_arr165[_454++] + (1U << CHAR_BIT)));
      _289 = data->dat_arr165[_454++];
      _289 += data->dat_arr165[_454++] << CHAR_BIT;
      fn224((int16_t)_289);
    } else
      write_stored_bits_to_buffer(data->dat_arr165[_454++]);
    if (data->uncompressible)
      return;
  }
  for (_226 = 0; _226 < CONST_N141_IS_511; _226++)
    data->dat_arr191[_226] = 0;
  for (_226 = 0; _226 < CONST_N142_IS_15; _226++)
    data->dat_arr193[_226] = 0;
}

File: archivelib-rs-tdd/src/compress/fn207.rs
use crate::compress::base::{CompressU16ArrayAlias, CompressU8ArrayAlias};
use crate::compress::{CompressError, RCompressData, Result};
use crate::consts::{
  CONST_N141_IS_511, CONST_N142_IS_15, CONST_N143_IS_9, CONST_N145_IS_19, CONST_N147_IS_5,
  CONST_N540_IS_5,
};
use crate::support::BitwiseWrite;
use std::io::Read;
const CHAR_BIT: usize = 8;

impl<R: Read, W: BitwiseWrite> RCompressData<R, W> {
  pub fn fn207(&mut self) -> Result<()> {
    let mut var456: u32 = 0 as u32;
    let mut var217 = [0; 2 * CONST_N145_IS_19 - 1];
    let mut var229 = self.fn211(
      CONST_N141_IS_511 as i32,
      &mut CompressU16ArrayAlias::Array191(0),
      &mut CompressU8ArrayAlias::Array180(0),
      &mut CompressU16ArrayAlias::Array192(0),
    )?;
    let var455 = self.dat_arr191[var229 as usize];
    self.output_store.write_bits(var455 as u32, 16 as usize)?;
    if var229 >= CONST_N141_IS_511 as u32 {
      self.fn216(&mut var217);
      var229 = self.fn211(
        CONST_N145_IS_19 as i32,
        &mut CompressU16ArrayAlias::Custom(0, &mut var217),
        &mut CompressU8ArrayAlias::Array181(0),
        &mut CompressU16ArrayAlias::Array194(0),
      )?;
      if var229 >= CONST_N145_IS_19 as u32 {
        self.fn218(CONST_N145_IS_19 as i16, CONST_N147_IS_5 as i16, 3 as i16)?;
      } else {
        self
          .output_store
          .write_bits(0u8, CONST_N147_IS_5 as usize)?;
        self
          .output_store
          .write_bits(var229 as u32, CONST_N147_IS_5 as usize)?;
      }
      self.fn222()?;
    } else {
      self
        .output_store
        .write_bits(0u8, CONST_N147_IS_5 as usize)?;
      self
        .output_store
        .write_bits(0u8, CONST_N147_IS_5 as usize)?;
      self
        .output_store
        .write_bits(0u8, CONST_N143_IS_9 as usize)?;
      self
        .output_store
        .write_bits(var229 as u32, CONST_N143_IS_9 as usize)?;
    }
    var229 = self.fn211(
      CONST_N142_IS_15 as i32,
      &mut CompressU16ArrayAlias::Array193(0),
      &mut CompressU8ArrayAlias::Array181(0),
      &mut CompressU16ArrayAlias::Array194(0),
    )?;
    if var229 >= CONST_N142_IS_15 as u32 {
      self.fn218(CONST_N142_IS_15 as i16, CONST_N540_IS_5 as i16, -1 as i16)?;
    } else {
      self
        .output_store
        .write_bits(0u8, CONST_N540_IS_5 as usize)?;
      self
        .output_store
        .write_bits(var229 as u32, CONST_N540_IS_5 as usize)?;
    }
    let mut var454 = 0 as u32;
    for run_start226 in 0..var455 {
      if run_start226 % 8 == 0 {
        var456 = self.dat_arr165[var454 as usize] as u32;
        var454 += 1;
      } else {
        var456 <<= 1;
      }
      if 0 != (var456 & (1 << (CHAR_BIT - 1))) {
        let val = self.dat_arr165[var454 as usize];
        self.write_stored_bits_to_buffer((val as i16).wrapping_add(1 << CHAR_BIT))?;
        var454 += 1;
        let var289 = self.dat_arr165[var454 as usize] as u32
          + ((self.dat_arr165[(var454 + 1) as usize] as u32) << 8);
        var454 += 2;
        self.fn224(var289 as u16)?;
      } else {
        let a1 = self.dat_arr165[var454 as usize] as i16;
        self.write_stored_bits_to_buffer(a1)?;
        var454 += 1;
      }
      if self.uncompressible {
        return Err(CompressError::InputUncompressable);
      }
    }
    for i in 0..CONST_N141_IS_511 {
      self.dat_arr191[i] = 0 as u16;
    }
    for i in 0..CONST_N142_IS_15 {
      self.dat_arr193[i] = 0 as u16;
    }
    Ok(())
  }
}
